# Embeddings:
             Are a way to represent words, phrases, or sentences as numbers

Imagine you have a magic tool that can turn any text into a set of numbers. These numbers are special because they capture the meaning and relationships between different words and sentences.

For example, let's say you have the words "cat" and "dog." The embeddings for these words might be two different sets of numbers. But since cats and dogs are both animals, the embeddings will be similar in some way. This similarity is measured by calculating the distance between the sets of numbers. If the distance is small, it means the words are related or similar. If the distance is large, it means the words are not very similar.

In [1]:
#Install the necessary libraries:
%pip install openai
%pip install tiktoken
%pip install pandas
%pip install python-dotenv
%pip install os


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement os (from versions: none)
ERROR: No matching distribution found for os


Uses of Embidding
The key idea behind embeddings is that words or sentences with similar meanings should have similar vector representations. By representing text inputs as vectors, various natural language processing tasks can be performed more effectively. For example, embeddings can be used for search engines to rank search results based on their relevance to a query. They can also be used for clustering, where similar text inputs are grouped together, or for recommendation systems to suggest items with related descriptions

# Import the necessary libraries:

In [1]:
import openai #The openai library provides access to the OpenAI API for generating embeddings.
from dotenv import load_dotenv,find_dotenv # Helps to load openai api key from dotenv file
import os #Helps to load openai api key from dotenv file
from tenacity import retry, wait_random_exponential, stop_after_attempt#The tenacity library helps in implementing retry logic to handle potential errors or failures when making API calls

 # Setup Openai Api

In [2]:
openai.api_key = os.environ.get("API_KEY")

# Generating the embedding:

In [ ]:
embedding = openai.Embedding.create(
    input=input("enter your text"), model="text-embedding-ada-002"
)["data"][0]["embedding"]


The openai.Embedding.create function is called to generate the embedding.
The model parameter specifies the model ID to use, in this case, "text-embedding-ada-002".
The API response is accessed using ["data"][0]["embedding"], which extracts the embedding from the response.
The obtained embedding is stored in the embedding variable.

# Checking the length of the embedding:

In [ ]:
print(len(embedding))

1536


 This line prints the length of the obtained embedding vector.
The length of the embedding vector indicates the number of dimensions or features captured in the embedding. In this case, the embedding model used (text-embedding-ada-002) produces embeddings of length

**This is the simple way to get the text converted in to numerical format let's see some practicle uses and examples with data**

# Embedding models

* OpenAI offers one second-generation embedding model (denoted by -002 in the model ID) and 16 first-generation models (denoted by -001 in the model ID).

We recommend using text-embedding-ada-002 for nearly all use cases. It’s better, cheaper, and simpler to use  

* MODEL GENERATION	TOKENIZER	MAX INPUT TOKENS	KNOWLEDGE CUTOFF
V2	cl100k_base	8191	Sep 2021
V1	GPT-2/GPT-3	2046	Aug 2020
Usage is priced per input token, at a rate of $0.0004 per 1000 tokens, or about ~3,000 pages per US dollar (assuming ~800 tokens per page):

MODEL	ROUGH PAGES PER DOLLAR	EXAMPLE PERFORMANCE ON BEIR SEARCH EVAL
text-embedding-ada-002	3000	53.9
*-davinci-*-001	6	52.8
*-curie-*-001	60	50.9
*-babbage-*-001	240	50.4
*-ada-*-001	300	49.0
Second-generation models
MODEL NAME	TOKENIZER	MAX INPUT TOKENS	OUTPUT DIMENSIONS
text-embedding-ada-002	cl100k_base	8191	1536